File: src/ai.ts
import { appConfig } from './config';
import { getRandomElement } from './util';
import { AIService, Provider } from './types';
import { UnauthorizedError, ValidationError } from './error';

/**
 * Reference
 *
 * https://github.com/Nutlope/aicommits/blob/develop/src/utils/prompt.ts
 *
 */
export const prompt = `Generate a single-line git commit message based on the provided information about staged and committed files, and the full diff. Adhere strictly to these specifications:
1. Format: <type>: <subject> OR <type>(<scope>): <subject>
   - <scope> is optional and should only be used when it adds significant clarity
2. Maximum length: 72 characters (including type and scope)
3. Use present tense and imperative mood
4. Capitalize the first letter of the subject
5. No period at the end
6. Message in only English language

Types:
- feat: New feature
- fix: Bug fix
- docs: Documentation changes
- style: Code style changes (formatting, missing semi colons, etc)
- refactor: Code refactoring
- perf: Performance improvements
- test: Adding or updating tests
- build: Build system or external dependency changes
- ci: CI configuration changes
- chore: Other changes that don't modify src or test files
- revert: Revert a previous commit

Guidelines:
- Be specific, concise, clear, and descriptive
- Focus on why the change was made, not how
- Use consistent terminology
- Avoid redundant information
- Analyze the file extensions to determine the appropriate type:
  - .ts, .tsx: TypeScript code (possibly React for .tsx)
  - .js, .jsx: JavaScript code (possibly React for .jsx)
  - .py: Python code
  - .go: Go code
  - .rb: Ruby code
  - .java: Java code
  - .cs: C# code
  - .cpp, .hpp, .h: C++ code
  - .md, .txt: Documentation
  - .yml, .yaml: Configuration files
  - .json: JSON data or configuration
  - .css, .scss, .less: Styling
  - .html: HTML markup
  - test.*, spec.*: Test files
- Consider both staged and committed files in determining the scope and nature of the change
- Analyze the full diff to understand the context and extent of the changes
- Only include scope when it significantly clarifies the change and fits within the character limit

Examples:
- feat(auth): Add user authentication feature
- fix(api): Resolve null pointer exception in login process
- docs: Update API endpoints documentation
- refactor(data): Simplify data processing algorithm
- test(utils): Add unit tests for string manipulation functions
- style: Format code according to style guide
- perf: Optimize database query for faster results

IMPORTANT: Respond ONLY with the commit message. Do not include any other text, explanations, or metadata. The entire response should be a single line containing only the commit message.`;

export const openAI: AIService = {
	generate: async (diff: string, apiKey?: string) => {
		try {
			const API_KEY = apiKey ? apiKey : appConfig.OPENAI_API_KEY;
			const response = await fetch('https://api.openai.com/v1/chat/completions', {
				method: 'POST',
				headers: {
					'Content-Type': 'application/json',
					Authorization: `Bearer ${API_KEY}`,
				},
				body: JSON.stringify({
					model: 'gpt-3.5-turbo',
					messages: [
						{
							role: 'system',
							content: prompt,
						},
						{
							role: 'user',
							content: diff,
						},
					],
					temperature: 0.7,
					max_tokens: 200,
				}),
			});

			if (!response.ok) {
				const errorData = await response.json();
				if (errorData.error?.code === 'invalid_api_key') {
					throw new UnauthorizedError(errorData.error.message);
				}
				if (errorData.error?.code === 'insufficient_quota') {
					throw new UnauthorizedError(
						'You exceeded your current quota, please check your plan and billing details',
					);
				}
				if (errorData.error?.code === 'context_length_exceeded') {
					throw new ValidationError('The provided input exceeds the maximum allowed token length.');
				}
				throw new Error(errorData.error?.message || 'Error calling OpenAI API');
			}

			const data = await response.json();
			const messages = data.choices
				.filter((choice: any) => choice.message?.content)
				.map((choice: any) => choice.message.content);
			return (getRandomElement(messages) as string)?.toLocaleLowerCase();
		} catch (error: any) {
			if (error instanceof UnauthorizedError || error instanceof ValidationError) {
				throw error;
			}
			throw error;
		}
	},
};

export const claudeAI: AIService = {
	generate: async (diff: string, apiKey?: string) => {
		try {
			const API_KEY = apiKey ? apiKey : appConfig.CLAUDE_API_KEY;
			const response = await fetch('https://api.anthropic.com/v1/messages', {
				method: 'POST',
				headers: {
					'Content-Type': 'application/json',
					'x-api-key': API_KEY,
					'anthropic-version': '2023-06-01',
				},
				body: JSON.stringify({
					model: 'claude-2.1',
					max_tokens: 1024,
					temperature: 0.7,
					messages: [
						{
							role: 'user',
							content: prompt + diff,
						},
					],
				}),
			});

			if (!response.ok) {
				const errorData = await response.json();
				if (errorData.type === 'authentication_error') {
					throw new UnauthorizedError(errorData.message);
				}
				throw new Error(errorData.message || 'Error calling Claude API');
			}

			const data = await response.json();
			const messages = data.content
				.filter((choice: any) => choice.text)
				.map((choice: any) => choice.text);
			return (getRandomElement(messages) as string)?.toLocaleLowerCase();
		} catch (error: any) {
			if (error instanceof UnauthorizedError) {
				throw error;
			}
			throw error;
		}
	},
};

export const deepseekAI: AIService = {
	generate: async (diff: string, apiKey?: string) => {
		try {
			const API_KEY = apiKey ? apiKey : appConfig.DEEPSEEK_API_KEY;
			const response = await fetch('https://api.deepseek.com/chat/completions', {
				method: 'POST',
				headers: {
					'Content-Type': 'application/json',
					Authorization: `Bearer ${API_KEY}`,
				},
				body: JSON.stringify({
					model: 'deepseek-chat',
					messages: [
						{
							role: 'system',
							content: prompt,
						},
						{
							role: 'user',
							content: diff,
						},
					],
					temperature: 0.7,
					max_tokens: 200,
				}),
			});

			if (!response.ok) {
				const errorData = await response.json();
				if (errorData.error?.type === 'invalid_api_key') {
					throw new UnauthorizedError(errorData.error.message);
				}
				throw new Error(errorData.error?.message || 'Error calling Deepseek API');
			}

			const data = await response.json();
			const messages = data.choices
				.filter((choice: any) => choice.message?.content)
				.map((choice: any) => choice.message.content);
			return (getRandomElement(messages) as string)?.toLocaleLowerCase();
		} catch (error: any) {
			if (error instanceof UnauthorizedError) {
				throw error;
			}
			throw error;
		}
	},
};

export const geminiAI: AIService = {
	generate: async (diff: string, apiKey?: string) => {
		try {
			const API_KEY = apiKey ? apiKey : appConfig.GEMINI_API_KEY;
			const response = await fetch(
				'https://generativelanguage.googleapis.com/v1beta/openai/chat/completions',
				{
					method: 'POST',
					headers: {
						'Content-Type': 'application/json',
						Authorization: `Bearer ${API_KEY}`,
					},
					body: JSON.stringify({
						model: 'gemini-2.0-flash',
						messages: [
							{
								role: 'system',
								content: prompt,
							},
							{
								role: 'user',
								content: diff,
							},
						],
						temperature: 0.7,
						max_tokens: 200,
					}),
				},
			);

			if (!response.ok) {
				const errorData = await response.json();
				if (errorData.error?.status === 'UNAUTHENTICATED') {
					throw new UnauthorizedError('Invalid API key or authentication error');
				}
				if (errorData.error?.status === 'RESOURCE_EXHAUSTED') {
					throw new UnauthorizedError(
						'You exceeded your current quota, please check your plan and billing details',
					);
				}
				throw new Error(errorData.error?.message || 'Error calling Gemini API');
			}

			const data = await response.json();
			const messages = data.choices
				.filter((choice: any) => choice.message?.content)
				.map((choice: any) => choice.message.content);
			return (getRandomElement(messages) as string)?.toLocaleLowerCase();
		} catch (error: any) {
			if (error instanceof UnauthorizedError) {
				throw error;
			}
			throw error;
		}
	},
};

export function ai(type?: Provider): AIService {
	switch (type) {
		case 'claudeai':
			return claudeAI;
		case 'deepseek':
			return deepseekAI;
		case 'gemini':
			return geminiAI;
		case 'openai':
		default:
			return openAI;
	}
}

File: cmd/ai.go
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

const prompt = `Write commit messages as:
<type>(<scope>): <short imperative intent>

Types:
- feat: new feature
- fix: bug fix
- docs: documentation changes
- style: format-only changes
- refactor: code restructuring without behavior change
- perf: performance improvements
- test: adding or updating tests
- build: build system or dependency changes
- ci: ci configuration changes
- chore: maintenance/tooling changes
- revert: revert a previous commit

Scope:
- include when it clarifies ownership: daemon, http, config, docs
- omit when obvious or repo-wide

Summary:
- lowercase
- imperative-ish
- no period
- describe intent, not file mechanics
- keep it under ~72 chars

Examples:
fix(daemon): retry failed deploys once per cycle
refactor(daemon): keep sync worker wiring in sync
test(daemon): cover same-commit reconciliation
docs: align daemon wiring map
chore(ci): pin node setup version`

type generateRequest struct {
	Diff            string
	DiffStat        string
	APIKey          string
	Suggestion      string
	PreviousMessage string
}

type generator interface {
	generate(req generateRequest) (string, error)
}

type chatMessage struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

type chatRequest struct {
	Model       string        `json:"model"`
	Messages    []chatMessage `json:"messages"`
	Temperature float64       `json:"temperature"`
	MaxTokens   int           `json:"max_tokens"`
}

type chatChoice struct {
	Message struct {
		Content string `json:"content"`
	} `json:"message"`
}

type chatResponse struct {
	Choices []chatChoice `json:"choices"`
}

type apiError struct {
	Error struct {
		Message string `json:"message"`
	} `json:"error"`
}

func buildMessages(diff, diffStat, suggestion, previousMessage string) []chatMessage {
	systemPrompt := prompt

	if strings.TrimSpace(suggestion) != "" && strings.TrimSpace(previousMessage) != "" {
		systemPrompt = fmt.Sprintf(`%s

The developer rejected this commit message: "%s"
The developer wants the commit message to: %s
Generate a completely new commit message that incorporates the developer's feedback. Still follow all formatting rules above.`, prompt, previousMessage, suggestion)
	}

	userContent := diff
	if strings.TrimSpace(diffStat) != "" {
		userContent = fmt.Sprintf("Summary of changed files (git diff --stat --summary):\n%s\n\nFull diff:\n%s", diffStat, diff)
	}

	return []chatMessage{
		{Role: "system", Content: systemPrompt},
		{Role: "user", Content: userContent},
	}
}

var httpClient = &http.Client{
	Timeout: 30 * time.Second,
}

func chatCompletion(apiURL, apiKey, model string, messages []chatMessage) (string, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	reqBody := chatRequest{
		Model:       model,
		Messages:    messages,
		Temperature: 0.2,
		MaxTokens:   200,
	}

	jsonData, err := json.Marshal(reqBody)
	if err != nil {
		return "", fmt.Errorf("marshaling request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, apiURL, bytes.NewBuffer(jsonData))
	if err != nil {
		return "", fmt.Errorf("creating request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+apiKey)

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return "", fmt.Errorf("sending request: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("reading response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return "", parseAPIError(body, resp.StatusCode)
	}

	var result chatResponse
	if err := json.Unmarshal(body, &result); err != nil {
		return "", fmt.Errorf("parsing response: %w", err)
	}

	if len(result.Choices) == 0 {
		return "", errors.New("no choices in api response")
	}

	return strings.TrimSpace(result.Choices[0].Message.Content), nil
}

func parseAPIError(body []byte, statusCode int) error {
	// try object format: {"error": {"message": "..."}}
	var objErr apiError
	if json.Unmarshal(body, &objErr) == nil && objErr.Error.Message != "" {
		return errors.New(objErr.Error.Message)
	}

	// try array format: [{"error": {"message": "..."}}]
	var arrErr []apiError
	if json.Unmarshal(body, &arrErr) == nil && len(arrErr) > 0 && arrErr[0].Error.Message != "" {
		return errors.New(arrErr[0].Error.Message)
	}

	return fmt.Errorf("api error: status %d", statusCode)
}

type openAI struct {
	config config
	url    string
}

func (s *openAI) generate(req generateRequest) (string, error) {
	apiKey := req.APIKey
	if strings.TrimSpace(apiKey) == "" {
		apiKey = s.config.openaiAPIKey
	}

	apiURL := s.url
	if apiURL == "" {
		apiURL = "https://api.openai.com/v1/chat/completions"
	}

	return chatCompletion(apiURL, apiKey, "gpt-3.5-turbo", buildMessages(req.Diff, req.DiffStat, req.Suggestion, req.PreviousMessage))
}

type gemini struct {
	config config
	url    string
}

func (s *gemini) generate(req generateRequest) (string, error) {
	apiKey := req.APIKey
	if strings.TrimSpace(apiKey) == "" {
		apiKey = s.config.geminiAPIKey
	}

	apiURL := s.url
	if apiURL == "" {
		apiURL = "https://generativelanguage.googleapis.com/v1beta/openai/chat/completions"
	}

	return chatCompletion(apiURL, apiKey, "gemini-2.5-flash-lite", buildMessages(req.Diff, req.DiffStat, req.Suggestion, req.PreviousMessage))
}

func ai(provider string, cfg config) generator {
	switch provider {
	case "openai":
		return &openAI{config: cfg}
	case "gemini":
		return &gemini{config: cfg}
	default:
		return &gemini{config: cfg}
	}
}
